java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitution.java

```diff
@@ -100,15 +100,13 @@ public TypeSubstitution combine(TypeSubstitution source) {
     TypeSubstitution result = new TypeSubstitution();
     for (Map.Entry<TypeVariableJavaType, JavaType> substitution : substitutionEntries()) {
       TypeVariableJavaType typeVar = substitution.getKey();
-      JavaType targetType = substitution.getValue();
-      if (targetType.isTagged(JavaType.WILDCARD)) {
-        targetType = ((WildCardType) targetType).bound;
-      }
+      JavaType targetType = filterWildcard(substitution.getValue());
       JavaType substitutedType = source.substitutedType(typeVar);
-      if(substitutedType == null) {
+      if(substitutedType == null || targetType == substitutedType) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)